### PR TITLE
[WIP] Improve rebuild time

### PR DESCRIPTION
**Pull request by pavelgrinchenko**
- [x] Save all necessary folders upon build
- [x] Remove outdated page-data folders

### CODE REVIEW

File: scripts/deploy-with-s3.js [hunk 1]
```diff
@@ -22,10 +22,15 @@ const path = require('path')
 const { execSync } = require('child_process')
 const { remove, move, ensureDir } = require('fs-extra')
 const { s3Prefix, s3Bucket, s3Client } = require('./s3-utils')
+const { cleanUpPageData } = require('./page-data-utils')
 
 const rootDir = path.join(__dirname, '..')
```
Comment (rogermparent) [inline scripts/deploy-with-s3.js:27]: I find `process.cwd()` is more solid than `__dirname` in getting a Gatsby project's root directory, but both methods do work.

File: scripts/page-data-utils.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    const itemPath = path.join(dir, item)
+    const itemStat = fs.statSync(itemPath)
+
+    if (itemStat.isDirectory()) {
+      if (!pagesTree[item]) {
+        console.warn(
+          `delete page-data folder for unexisted page at ${itemPath}`
+        )
+        return deleteDir(itemPath)
+        return
```
Comment (rogermparent) [inline scripts/page-data-utils.js:68]: This second return could be a typo

File: scripts/page-data-utils.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+  pagesParts.forEach(pageParts => handlePageParts(tree, pageParts))
+  exceptionFolders.forEach(exception => (tree[exception] = {}))
+  return tree
+}
+
+const deleteDir = dir => {
+  if (fs.existsSync(dir)) {
+    fs.readdirSync(dir).forEach(item => {
+      const currentPath = path.join(path, item)
```
Comment (rogermparent) [inline scripts/page-data-utils.js:44]: Did you mean `path.join(dir, item)`? This breaks when it's called.

File: scripts/deploy-with-s3.js [hunk 4]
```diff
@@ -112,16 +130,23 @@ async function main() {
...
+  cleanUpPageData()
+  console.timeEnd('Cleaned up outdated page data files for')
+
+  await move(
+    path.join(localPath(publicDirName), '404.html'),
+    path.join(rootDir, '404.html'),
+    {
+      overwrite: true
+    }
+  )
```
Comment (rogermparent) [inline scripts/deploy-with-s3.js:147]: I'm not sure what moving the 404 page from `public` to the project root does.

**Comment by shcheklein** [comment]
Closing in favor of https://github.com/iterative/dvc.org/pull/1182 ... we making a step back a bit, there is an explanation in that PR @rogermparent provided.

**Pull request closed without merging**